### PR TITLE
[Notify] Add Notify bundle

### PR DESCRIPTION
**Pull request by mtarld**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| Tickets       | 
| License       | MIT

In Symfony 5.3, Notifier was shipped with a Mercure bridge (https://github.com/symfony/symfony/pull/39342)

Then, with a simple `$this->notifier->send(new Notification('My message', ['chat/mercure']));` (and a running Mercure server), it'll be easy to create "Server-Sent-Events".

Therefore, the Notify bundle idea is to listen to these events using JavaScript event sourcing and convert them as native HTML5 notifications (using the `{{ stream_notifications() }}` Twig function).

(The "Notify" name was the first that came to my mind a could/should be challenged)

### CODE REVIEW

File: src/Notify/Resources/assets/src/controller.js
Comment (dunglas) [inline src/Notify/Resources/assets/src/controller.js:19]: Shouldn't you use [values](https://stimulus.hotwire.dev/reference/values) instead?

File: src/Notify/Resources/assets/src/controller.js
Comment (dunglas) [inline src/Notify/Resources/assets/src/controller.js:23]: Is it a good idea to dispatch this event when it's not possible to connect like in this case?

**Review by dunglas — CHANGES_REQUESTED** [review]

File: src/Notify/Resources/assets/src/controller.js
Comment (mtarld) [reply src/Notify/Resources/assets/src/controller.js:19]: Yes indeed! Thanks

File: src/Notify/Twig/NotifyExtension.php
Comment (mtarld) [reply src/Notify/Twig/NotifyExtension.php:53]: As discussed together, it does make sense but it has to be more meaningful. Therefore I renamed `{{ notify() }}` to `{{ stream_notifications() }}`

**Comment by seb-jean** [comment]
Hi 😃 ,

With this bundle, will it be web push notifications (with allow and block) or simple notifications?

Push notifications :
Ca1pture

notification-message

Simple notification :
Capture

Thanks :)

**Comment by tgalopin** [comment]
Push notifications (native)

**Comment by tgalopin** [comment]
Hello @mtarld 

I think this component still makes sense, could we perhaps use latest Symfony changes (especially on ux-mercure) to finalize it? Would you have time to have a look at the failing tests/rebasing?

**Comment by mtarld** [comment]
Hi @tgalopin! 
Yes indeed, I think it is worth it! A lot of improvements and changes happened, I'll update the bundle consequently :slightly_smiling_face:

**Comment by mtarld** [comment]
Hi @tgalopin! I updated the bundle to be consistent with other bundles (and to reuse some good ideas taken from ux-turbo-mercure)

Still, some questions remain: 
- As the notify bundle is tightly coupled to mercure, should we name the bundle after it? Something like `MercureNotify` with a `stream_mercure_notification` twig function?
- Actually, `symfony/mercure-notifier` is a strong composer dependency of the bundle. But it could actually work without (as the bundle is just a "listener"). Maybe we can instead change that dependency to a composer suggestion?

File: src/Notify/README.md
Comment (yceruto) [inline src/Notify/README.md]: If this is a common need, I'd suggest creating a config for it, so one could change easily the Mercure hub linked to this service:
```yaml
notify:
    mercure_hub: mercure.hub.default # default service id
```

Comment (mtarld) [reply src/Notify/README.md]: That's a great idea indeed! :+1:

File: src/Notify/DependencyInjection/NotifyExtension.php
Comment (yceruto) [inline src/Notify/DependencyInjection/NotifyExtension.php]: Now that Configuration is added, you can use `ConfigurableExtension` for code simplification.

Comment (mtarld) [reply src/Notify/DependencyInjection/NotifyExtension.php]: Didn't know about it! Many thanks

File: src/Notify/DependencyInjection/NotifyExtension.php
Comment (yceruto) [inline src/Notify/DependencyInjection/NotifyExtension.php]: you can omit this definition, it's false by default

File: src/Notify/DependencyInjection/NotifyExtension.php
Comment (yceruto) [inline src/Notify/DependencyInjection/NotifyExtension.php]: should we ask for ` class_exists(StimulusTwigExtension::class)` too?

**Review by yceruto — APPROVED** [review]
LGTM

**Review by dunglas — APPROVED** [review]

File: src/Notify/Resources/doc/index.rst
Comment (kbond) [inline src/Notify/Resources/doc/index.rst]:
```suggestion

```

Comment (mtarld) [reply src/Notify/Resources/doc/index.rst]: Oops! Nice catch thanks!

File: src/Notify/Twig/NotifyExtension.php
Comment (chalasr) [inline src/Notify/Twig/NotifyExtension.php]: let's leverage PHP 8.1+ first-class callables

```suggestion
        yield new TwigFunction('stream_notifications', $this->renderStreamNotifications(...), ['needs_environment' => true, 'is_safe' => ['html']]);
```

File: src/Notify/Twig/NotifyExtension.php
Comment (chalasr) [inline src/Notify/Twig/NotifyExtension.php]:
```suggestion
    private function renderStreamNotifications(Environment $environment, array|string $topics = [], array $options = []): string
    {
```

File: src/Notify/Resources/doc/index.rst
Comment (weaverryan) [inline src/Notify/Resources/doc/index.rst]: I'd move this to the last section, as most people will not actually need to configure this. And if you DO have multiple hubs, you will be looking specifically for this config.

File: src/Notify/Resources/doc/index.rst
Comment (weaverryan) [inline src/Notify/Resources/doc/index.rst:52]: We should include the `#[AsCommand]` above this, just to make this a complete example.

File: src/Notify/Twig/NotifyExtension.php
Comment (weaverryan) [reply src/Notify/Twig/NotifyExtension.php]: ~~Should we? I'd like to keep php 8.0 compat at least~~ Nope: we will require 8.1

File: src/Notify/composer.json
Comment (weaverryan) [inline src/Notify/composer.json]: I think, for new stuff, we should require php 8.0 at least. But I've asked the core team to weigh in

UPDATE: Require 8.1 as the minimum. And then, feel free to use any fancy 8.1 features you want ;)

File: src/Notify/composer.json
Comment (weaverryan) [inline src/Notify/composer.json]: For Symfony packages, make 5.4 the minimum.

File: src/Notify/Twig/NotifyExtension.php
Comment (weaverryan) [inline src/Notify/Twig/NotifyExtension.php]: Hmm, we should probably leverage a Twig extension "runtime class" - https://symfony.com/doc/current/templating/twig_extension.html#creating-lazy-loaded-twig-extensions - so that these dependencies (especially the hub) aren't loaded unless the user actually calls this Twig function.

File: src/Notify/Resources/assets/src/controller.ts
Comment (weaverryan) [inline src/Notify/Resources/assets/src/controller.ts]: Also add `bubbles: true` as a sibling to `detail`. This is something we have done incorrectly in other packages so far.

File: src/Notify/Resources/doc/index.rst
Comment (weaverryan) [inline src/Notify/Resources/doc/index.rst]: I'd love a screenshot of a notification. I didn't really understand what you meant by "native notification" until I really dug into the code. Maybe also link `native notifications` to https://developer.mozilla.org/en-US/docs/Web/API/Notifications_API/Using_the_Notifications_API

File: src/Notify/README.md
Comment (weaverryan) [inline src/Notify/README.md]: And also here, link `native notifications` to https://developer.mozilla.org/en-US/docs/Web/API/Notifications_API/Using_the_Notifications_API

File: src/Notify/Resources/doc/index.rst
Comment (weaverryan) [inline src/Notify/Resources/doc/index.rst]:
```rst
Finally, to "listen" and trigger the notifications in the user's browser, call the ``stream_notifications()`` Twig function anywhere on the page:
```

File: src/Notify/Resources/doc/index.rst
Comment (weaverryan) [inline src/Notify/Resources/doc/index.rst]:
```suggestion
   Calling ``stream_notifications()`` without parameter will fallback to the
```

File: src/Notify/Resources/doc/index.rst
Comment (weaverryan) [inline src/Notify/Resources/doc/index.rst:72]: I'm missing "the big picture" a little bit. Here is my understanding:

A) I send a `Notification` in PHP to the `chat/mercure` channel.
B) This communicates with https://github.com/symfony/mercure-notifier to send that notification to my Mercure hub
C) ??? How/where do the Mercure topics "connect" with this? If I listen to the `/my/topic/1` topic here... which notifications would I receive? Is the "topic" configured somewhere when sending the `Notification` to mercure?

File: src/Notify/Resources/doc/index.rst
Comment (weaverryan) [inline src/Notify/Resources/doc/index.rst]:
```suggestion
Extend the Stimulus Controller
```

might be better

File: src/Notify/Resources/doc/index.rst
Comment (weaverryan) [inline src/Notify/Resources/doc/index.rst]:
```suggestion
    // assets/controllers/mynotify_controller.js
```

File: src/Notify/Resources/doc/index.rst
Comment (weaverryan) [inline src/Notify/Resources/doc/index.rst:105]: I hate to have to worry about it, but with this example, users are going to have the `this` problem down in `_onConnect()`. So perhaps we should add:

```js
initialize() {
    // guarantees "this" refers to this object in _onConnect
    this._onConnect = this._onConnect.bind(this);
}
```

File: src/Notify/DependencyInjection/NotifyExtension.php
Comment (weaverryan) [inline src/Notify/DependencyInjection/NotifyExtension.php]: Why not make these "hard" dependencies in `composer.json`? My thinking: the main/only thing this bundle provides is the `stream_notifications()` Twig function. If you install this package and are missing either or these two dependencies (especially `StimulusTwigExtension`), it could confuse people to silently not register the new function. So, I think we make these both real dependencies.

(side note, you could argue that Encore shouldn't be needed... just to get the `StimulusTwigExtension` and that's valid. But that's a separate topic).

**Review by weaverryan — COMMENTED** [review]
Very cool 😎

File: src/Notify/Resources/doc/index.rst
Comment (mtarld) [reply src/Notify/Resources/doc/index.rst:72]: When you send a notification through the mercure notifier, you'll have to specify channels in the `Notification`:
```php
$this->notifier->send(new Notification('Hello native notifications!', ['chat/mercureFoo']));
```

This channel is configured in the notifier configuration, there you can set a dsn:
```yaml
framework:
    notifier:
        chatter_transports:
           mercureFoo: '%env(MERCURE_FOO_DSN)%'
```

And it that very dsn, you'll be able to configure the topics to send the notification:
```
MERCURE_FOO_DSN=mercure://default?topics[]=fooOne&topics[]=fooTwo
MERCURE_BAR_DSN=mercure://default?topics[]=bar
```

Finally, you can subscribe the topics you want using the twig helper:
```twig
{{ stream_notifications(['fooOne']) }}
```

Is it now clearer for you? Maybe it worth to be written somewhere in the documentation?

File: src/Notify/DependencyInjection/NotifyExtension.php
Comment (mtarld) [reply src/Notify/DependencyInjection/NotifyExtension.php]: 100% agreed!

**Comment by mtarld** [comment]
As we required PHP 8.1 for that bundle, I updated the test workflow to run the bundle's tests on PHP 8.1.
I think we may test other bundles over PHP 8.1 instead of 8.0 in another PR. 
WDYT?

**Comment by mtarld** [comment]
Thanks for that wonderful review @weaverryan :slightly_smiling_face:!

File: src/Notify/Resources/doc/index.rst
Comment (weaverryan) [reply src/Notify/Resources/doc/index.rst:72]: Yes, you explained that VERY clearly. I understand that, technically, much of this should be documented in the notify + mercure bridge. But yes, it'd be great to include this in the README. I had originally thought that the `chat/mercure` channel used in your example was a special string that NEEDED to be exactly that string. Clearly, I was missing the point :)

**Comment by weaverryan** [comment]
> 
As we required PHP 8.1 for that bundle, I updated the test workflow to run the bundle's tests on PHP 8.1.
I think we may test other bundles over PHP 8.1 instead of 8.0 in another PR.
WDYT?

The test matrix DOES need some work. But if a bundle advertises that it works on php8, we need to keep testing it on php 8.0 (though, we should probably also test on php 8.1). Btw, I think we need to improve the test matrix further to be dynamic. There is way too much repeated config, and that's only going to get worse as the number of components grows. If you're interested, I had a WIP version of this at https://github.com/weaverryan/ux/tree/typescript-central-ci - where each package created a dynamic build. I think it's an interesting approach.

For this PR, I'd love to see your excellent explanation put into the docs. But after that, it's ready to merge!

**Review by maxhelias — APPROVED** [review]

File: src/Notify/Resources/doc/index.rst
Comment (weaverryan) [inline src/Notify/Resources/doc/index.rst:45]:
```suggestion
Then, you can inject the ``NotifierInterface`` service and send messages on the ``chat/myMercureChatter `` channel::
```

File: src/Notify/Resources/doc/index.rst
Comment (weaverryan) [inline src/Notify/Resources/doc/index.rst:87]: This looks a tiny bit to short

**Review by weaverryan — APPROVED** [review]

**Comment by weaverryan** [comment]
Thank you Mathias! Wonderful work!

**Comment by weaverryan** [comment]
Docs tweak over at #325

**Pull request merged**
